biz.aQute.bndlib/src/aQute/bnd/osgi/Macro.java

```diff
@@ -607,16 +609,22 @@ public String _replace(String args[]) {
 		return sb.toString();
 	}
 
-	public String _warning(String args[]) {
+	public String _warning(String args[]) throws Exception {
 		for (int i = 1; i < args.length; i++) {
-			domain.warning("%s", process(args[i]));
+			SetLocation warning = domain.warning("%s", process(args[i]));
+			FileLine header = domain.getHeader(Pattern.compile(".*"), Pattern.compile("\\$\\{warning;"));
```

Comment:
Would it make sense to put the patterns into `private static final` variables to avoid repeated pattern compilation?